clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -908,311 +628,781 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
+                              Handlers &... handlers) {
+#define KF_FOR_EACH(FUNC)                                                      \
+  (void)std::initializer_list<int> { (handlers.FUNC(Field, FieldTy), 0)... }
+
+  for (const auto &Field : Fields) {
+    QualType FieldTy = Field->getType();
+
+    if (Util::isSyclAccessorType(FieldTy))
+      KF_FOR_EACH(handleSyclAccessorType);
+    else if (Util::isSyclSamplerType(FieldTy))
+      KF_FOR_EACH(handleSyclSamplerType);
+    else if (Util::isSyclSpecConstantType(FieldTy))
+      KF_FOR_EACH(handleSyclSpecConstantType);
+    else if (Util::isSyclStreamType(FieldTy)) {
+      // Stream actually wraps accessors, so do recursion
+      CXXRecordDecl *RD = FieldTy->getAsCXXRecordDecl();
+      VisitAccessorWrapper(nullptr, Field, RD, handlers...);
+      KF_FOR_EACH(handleSyclStreamType);
+    } else if (FieldTy->isStructureOrClassType()) {
+      KF_FOR_EACH(handleStructType);
+      CXXRecordDecl *RD = FieldTy->getAsCXXRecordDecl();
+      VisitAccessorWrapper(nullptr, Field, RD, handlers...);
+    } else if (FieldTy->isReferenceType())
+      KF_FOR_EACH(handleReferenceType);
+    else if (FieldTy->isPointerType())
+      KF_FOR_EACH(handlePointerType);
+    else if (FieldTy->isArrayType())
+      KF_FOR_EACH(handleArrayType);
+    else if (FieldTy->isScalarType())
+      KF_FOR_EACH(handleScalarType);
+    else
+      KF_FOR_EACH(handleOtherType);
+  }
+#undef KF_FOR_EACH
+}
+
+// A base type that the SYCL OpenCL Kernel construction task uses to implement
+// individual tasks.
+template <typename Derived> class SyclKernelFieldHandler {
+protected:
+  Sema &SemaRef;
+  SyclKernelFieldHandler(Sema &S) : SemaRef(S) {}
+
+public:
+  // Mark these virutal so that we can use override in the implementer classes,
+  // despite virtual dispatch never being used.
+
+  //// TODO: Can these return 'bool' and we can short-circuit the handling? That
+  // way the field checker cna return true/false based on whether the rest
+  // should be still working.
+
+  // Accessor can be a base class or a field decl, so both must be handled.
+  virtual void handleSyclAccessorType(const CXXBaseSpecifier &, QualType) {}
+  virtual void handleSyclAccessorType(FieldDecl *, QualType) {}
+  virtual void handleSyclSamplerType(FieldDecl *, QualType) {}
+  virtual void handleSyclSpecConstantType(FieldDecl *, QualType) {}
+  virtual void handleSyclStreamType(const CXXBaseSpecifier &, QualType) {}
+  virtual void handleSyclStreamType(FieldDecl *, QualType) {}
+  virtual void handleStructType(FieldDecl *, QualType) {}
+  virtual void handleReferenceType(FieldDecl *, QualType) {}
+  virtual void handlePointerType(FieldDecl *, QualType) {}
+  virtual void handleArrayType(FieldDecl *, QualType) {}
+  virtual void handleScalarType(FieldDecl *, QualType) {}
+  // Most handlers shouldn't be handling this, just the field checker.
+  virtual void handleOtherType(const FieldDecl *, QualType) {}
```

Comment:
I think we need to remove the 'const' here as well, just to make the interface uniform.